---
title: "NIDDK-15 CellChat Analysis"
subtitle: "Trimester `r params$timepoint`: Without vs `r params$group %>% str_to_title()`"
author: "Vicky Chen"
date: "`r Sys.Date()`"
output:
    rmdformats::robobook:
        lightbox: true
        number_sections: true
        code-fold: true
        toc_depth: 3
        fig_width: 8
        fig_height: 8
params:
  timepoint: timepoint
  group: group
  path: path
---

```{r setup, include=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = FALSE, include=TRUE, message=FALSE, warning=FALSE, error=FALSE)
```

```{r libraries}
library(CellChat)
library(Seurat)
library(patchwork)
library(ComplexHeatmap)
library(DT)
library(knitr)
library(stringr)
options(stringsAsFactors = FALSE)
```

```{r paths}
seurat_obj <- "/data/NHLBI_IDSS/rawdata/NIDDK-15_temp/count/merge_processing/harmony/timepoint_patient/analysis1/analysis1_subset.rds"
# rename B-cells
b_cells <- read.csv("/data/NHLBI_IDSS/rawdata/NIDDK-15_temp/count/merge_processing/seurat_merge/B_cells/Bcells_1Tri_FCRL1_metadata.csv", row.names = 1)
#cellchat_path <- paste0("/data/NHLBI_IDSS/rawdata/NIDDK-15_temp/cellchat_BD/trimester", params$timepoint)
cellchat_path <- file.path(params$path, paste0("trimester", params$timepoint))
index <- list(after=1, before=2, uninfected=3, without=4)
comparison <- c(index[[params$group]], 4) # 1 = after, 2 = before, 3 = uninfected, 4 = without 
```

```{r settings}
colors = list(`1`='#337AB7', `2`='#6AA84F', `3`='#A70505')
```

```{r redefine function}
netVisual_circle_strength <- function (net, color.use = NULL, title.name = NULL, sources.use = NULL,
    targets.use = NULL, idents.use = NULL, remove.isolate = FALSE,
    top = 1, weight.scale = FALSE, vertex.weight = 20, vertex.weight.max = NULL,
    vertex.size.max = NULL, vertex.label.cex = 1, vertex.label.color = "black",
    edge.weight.max = NULL, edge.width.max = 8, alpha.edge = 0.6,
    label.edge = FALSE, edge.label.color = "black", edge.label.cex = 0.8,
    edge.curved = 0.2, shape = "circle", layout = in_circle(),
    margin = 0.2, vertex.size = NULL, arrow.width = 1, arrow.size = 0.2,
    text.x = 0, text.y = 1.5)
{
    if (!is.null(vertex.size)) {
        warning("'vertex.size' is deprecated. Use `vertex.weight`")
    }
    if (is.null(vertex.size.max)) {
        if (length(unique(vertex.weight)) == 1) {
            vertex.size.max <- 5
        }
        else {
            vertex.size.max <- 15
        }
    }
    options(warn = -1)
    thresh <- stats::quantile(net, probs = 1 - top)
    net[net < thresh] <- 0
    if ((!is.null(sources.use)) | (!is.null(targets.use)) | (!is.null(idents.use))) {
        if (is.null(rownames(net))) {
            stop("The input weighted matrix should have rownames!")
        }
        cells.level <- rownames(net)
        df.net <- reshape2::melt(net, value.name = "value")
        colnames(df.net)[1:2] <- c("source", "target")
        if (!is.null(sources.use)) {
            if (is.numeric(sources.use)) {
                sources.use <- cells.level[sources.use]
            }
            df.net <- subset(df.net, source %in% sources.use)
        }
        if (!is.null(targets.use)) {
            if (is.numeric(targets.use)) {
                targets.use <- cells.level[targets.use]
            }
            df.net <- subset(df.net, target %in% targets.use)
        }
        if (!is.null(idents.use)) {
            if (is.numeric(idents.use)) {
                idents.use <- cells.level[idents.use]
            }
            df.net <- filter(df.net, (source %in% idents.use) |
                (target %in% idents.use))
        }
        df.net$source <- factor(df.net$source, levels = cells.level)
        df.net$target <- factor(df.net$target, levels = cells.level)
        df.net$value[is.na(df.net$value)] <- 0
        net <- tapply(df.net[["value"]], list(df.net[["source"]],
            df.net[["target"]]), sum)
    }
    net[is.na(net)] <- 0
    if (remove.isolate) {
        idx1 <- which(Matrix::rowSums(net) == 0)
        idx2 <- which(Matrix::colSums(net) == 0)
        idx <- intersect(idx1, idx2)
        net <- net[-idx, ]
        net <- net[, -idx]
    }
    g <- graph_from_adjacency_matrix(net, mode = "directed",
        weighted = T)
    edge.start <- igraph::ends(g, es = igraph::E(g), names = FALSE)
    coords <- layout_(g, layout)
    if (nrow(coords) != 1) {
        coords_scale = scale(coords)
    }
    else {
        coords_scale <- coords
    }
    if (is.null(color.use)) {
        color.use = scPalette(length(igraph::V(g)))
    }
    if (is.null(vertex.weight.max)) {
        vertex.weight.max <- max(vertex.weight)
    }
    vertex.weight <- vertex.weight/vertex.weight.max * vertex.size.max +
        5
    loop.angle <- ifelse(coords_scale[igraph::V(g), 1] > 0, -atan(coords_scale[igraph::V(g),
        2]/coords_scale[igraph::V(g), 1]), pi - atan(coords_scale[igraph::V(g),
        2]/coords_scale[igraph::V(g), 1]))
    igraph::V(g)$size <- vertex.weight
    igraph::V(g)$color <- color.use[igraph::V(g)]
    igraph::V(g)$frame.color <- color.use[igraph::V(g)]
    igraph::V(g)$label.color <- vertex.label.color
    igraph::V(g)$label.cex <- vertex.label.cex
    if (label.edge) {
        igraph::E(g)$label <- igraph::E(g)$weight
        igraph::E(g)$label <- round(igraph::E(g)$label, digits = 2)
    }
    if (is.null(edge.weight.max)) {
        edge.weight.max <- max(igraph::E(g)$weight)
    }
    if (weight.scale == TRUE) {
        igraph::E(g)$width <- 0.3 + igraph::E(g)$weight/edge.weight.max *
            edge.width.max
    }
    else {
        igraph::E(g)$width <- 0.3 + edge.width.max * igraph::E(g)$weight
    }
    igraph::E(g)$arrow.width <- arrow.width
    igraph::E(g)$arrow.size <- arrow.size
    igraph::E(g)$label.color <- edge.label.color
    igraph::E(g)$label.cex <- edge.label.cex
    igraph::E(g)$color <- grDevices::adjustcolor(igraph::V(g)$color[edge.start[,
        1]], alpha.edge)
    igraph::E(g)$loop.angle <- rep(0, length(igraph::E(g)))
    if (sum(edge.start[, 2] == edge.start[, 1]) != 0) {
        igraph::E(g)$loop.angle[which(edge.start[, 2] == edge.start[,
            1])] <- loop.angle[edge.start[which(edge.start[,
            2] == edge.start[, 1]), 1]]
    }
    radian.rescale <- function(x, start = 0, direction = 1) {
        c.rotate <- function(x) (x + start)%%(2 * pi) * direction
        c.rotate(scales::rescale(x, c(0, 2 * pi), range(x)))
    }
    label.locs <- radian.rescale(x = 1:length(igraph::V(g)),
        direction = -1, start = 0)
    label.dist <- vertex.weight/max(vertex.weight) + 2
    plot(g, edge.curved = edge.curved, vertex.shape = shape,
        layout = coords_scale, margin = margin, vertex.label.dist = label.dist,
        vertex.label.degree = label.locs, vertex.label.family = "Helvetica",
        edge.label.family = "Helvetica")
    if (!is.null(title.name)) {
        text(text.x, text.y, title.name, cex = 1.1)
    }
    gg <- recordPlot()
    return(gg)
}

netVisual_diffInteraction_strength <- function (object, comparison = c(1, 2), measure = c("count",
    "weight", "count.merged", "weight.merged"), color.use = NULL,
    color.edge = c("#b2182b", "#2166ac"), title.name = NULL,
    sources.use = NULL, targets.use = NULL, remove.isolate = FALSE,
    top = 1, weight.scale = FALSE, vertex.weight = 20, vertex.weight.max = NULL,
    vertex.size.max = 15, vertex.label.cex = 1, vertex.label.color = "black",
    edge.weight.max = NULL, edge.width.max = 8, alpha.edge = 0.6,
    label.edge = FALSE, edge.label.color = "black", edge.label.cex = 0.8,
    edge.curved = 0.2, shape = "circle", layout = in_circle(),
    margin = 0.2, arrow.width = 1, arrow.size = 0.2)
{
    options(warn = -1)
    measure <- match.arg(measure)
    obj1 <- object@net[[comparison[1]]][[measure]]
    obj2 <- object@net[[comparison[2]]][[measure]]
    net.diff <- obj2 - obj1
    if (measure %in% c("count", "count.merged")) {
        if (is.null(title.name)) {
            title.name = "Differential number of interactions"
        }
    }
    else if (measure %in% c("weight", "weight.merged")) {
        if (is.null(title.name)) {
            title.name = "Differential interaction strength"
        }
    }
    net <- net.diff
    if ((!is.null(sources.use)) | (!is.null(targets.use))) {
        df.net <- reshape2::melt(net, value.name = "value")
        colnames(df.net)[1:2] <- c("source", "target")
        if (!is.null(sources.use)) {
            if (is.numeric(sources.use)) {
                sources.use <- rownames(net.diff)[sources.use]
            }
            df.net <- subset(df.net, source %in% sources.use)
        }
        if (!is.null(targets.use)) {
            if (is.numeric(targets.use)) {
                targets.use <- rownames(net.diff)[targets.use]
            }
            df.net <- subset(df.net, target %in% targets.use)
        }
        cells.level <- rownames(net.diff)
        df.net$source <- factor(df.net$source, levels = cells.level)
        df.net$target <- factor(df.net$target, levels = cells.level)
        df.net$value[is.na(df.net$value)] <- 0
        net <- tapply(df.net[["value"]], list(df.net[["source"]],
            df.net[["target"]]), sum)
        net[is.na(net)] <- 0
    }
    if (remove.isolate) {
        idx1 <- which(Matrix::rowSums(net) == 0)
        idx2 <- which(Matrix::colSums(net) == 0)
        idx <- intersect(idx1, idx2)
        net <- net[-idx, ]
        net <- net[, -idx]
    }
    net[abs(net) < stats::quantile(abs(net), probs = 1 - top,
        na.rm = T)] <- 0
    g <- graph_from_adjacency_matrix(net, mode = "directed",
        weighted = T)
    edge.start <- igraph::ends(g, es = igraph::E(g), names = FALSE)
    coords <- layout_(g, layout)
    if (nrow(coords) != 1) {
        coords_scale = scale(coords)
    }
    else {
        coords_scale <- coords
    }
    if (is.null(color.use)) {
        color.use = scPalette(length(igraph::V(g)))
    }
    if (is.null(vertex.weight.max)) {
        vertex.weight.max <- max(vertex.weight)
    }
    vertex.weight <- vertex.weight/vertex.weight.max * vertex.size.max +
        5
    loop.angle <- ifelse(coords_scale[igraph::V(g), 1] > 0, -atan(coords_scale[igraph::V(g),
        2]/coords_scale[igraph::V(g), 1]), pi - atan(coords_scale[igraph::V(g),
        2]/coords_scale[igraph::V(g), 1]))
    igraph::V(g)$size <- vertex.weight
    igraph::V(g)$color <- color.use[igraph::V(g)]
    igraph::V(g)$frame.color <- color.use[igraph::V(g)]
    igraph::V(g)$label.color <- vertex.label.color
    igraph::V(g)$label.cex <- vertex.label.cex
    if (label.edge) {
        igraph::E(g)$label <- igraph::E(g)$weight
        igraph::E(g)$label <- round(igraph::E(g)$label, digits = 2)
    }
    igraph::E(g)$arrow.width <- arrow.width
    igraph::E(g)$arrow.size <- arrow.size
    igraph::E(g)$label.color <- edge.label.color
    igraph::E(g)$label.cex <- edge.label.cex
    igraph::E(g)$color <- ifelse(igraph::E(g)$weight > 0, color.edge[1],
        color.edge[2])
    igraph::E(g)$color <- grDevices::adjustcolor(igraph::E(g)$color,
        alpha.edge)
    igraph::E(g)$weight <- abs(igraph::E(g)$weight)
    if (is.null(edge.weight.max)) {
        edge.weight.max <- max(igraph::E(g)$weight)
    }
    if (weight.scale == TRUE) {
        igraph::E(g)$width <- 0.3 + igraph::E(g)$weight/edge.weight.max *
            edge.width.max
    }
    else {
        igraph::E(g)$width <- 0.3 + edge.width.max * igraph::E(g)$weight
    }
    igraph::E(g)$loop.angle <- 0
    if (sum(edge.start[, 2] == edge.start[, 1]) != 0) {
        igraph::E(g)$loop.angle[which(edge.start[, 2] == edge.start[,
            1])] <- loop.angle[edge.start[which(edge.start[,
            2] == edge.start[, 1]), 1]]
    }
    radian.rescale <- function(x, start = 0, direction = 1) {
        c.rotate <- function(x) (x + start)%%(2 * pi) * direction
        c.rotate(scales::rescale(x, c(0, 2 * pi), range(x)))
    }
    label.locs <- radian.rescale(x = 1:length(igraph::V(g)),
        direction = -1, start = 0)
    label.dist <- vertex.weight/max(vertex.weight) + 2
    plot(g, edge.curved = edge.curved, vertex.shape = shape,
        layout = coords_scale, margin = margin, vertex.label.dist = label.dist,
        vertex.label.degree = label.locs, vertex.label.family = "Helvetica",
        edge.label.family = "Helvetica")
    if (!is.null(title.name)) {
        text(0, 1.5, title.name, cex = 1.1)
    }
    gg <- recordPlot()
    return(gg)
}
```

<!-- # Annotated UMAPs {.tabset .tabset-fade} -->
<!-- ## Cell-Type Grouped {-} -->
<!-- ```{r annotated UMAP} -->
<!-- seur <- readRDS(seurat_obj) -->

<!-- seur <- AddMetaData(seur, b_cells) -->
<!-- seur@meta.data <- seur@meta.data %>% -->
<!--   mutate(predicted.celltype.l1.split = coalesce(FCRL1, predicted.celltype.l1)) -->

<!-- # create umap of predicted.ids -->
<!-- DimPlot(seur, reduction = 'umap', group.by = "predicted.celltype.l1.split") -->
<!-- ``` -->

<!-- ## Cell-Type Split {-} -->
<!-- ```{r split annotated UMAP} -->
<!-- # create umap of split on flare -->
<!-- DimPlot(seur, reduction = 'umap', split.by = "cr.flare", group.by = "predicted.celltype.l1.split", ncol = 2) -->
<!-- ``` -->

<!-- ## Clusters {-} -->
<!-- ```{r Cluster UMAP} -->
<!-- # create umap of split on flare -->
<!-- DimPlot(seur, reduction = 'umap') -->
<!-- ``` -->

<!-- ## Cell Counts {-} -->
<!-- ```{r cell counts} -->
<!-- table(seur@meta.data$predicted.celltype.l1.split, seur@meta.data$cr.flare) -->
<!-- ``` -->

<!-- ```{r cleanup, results="hide"} -->
<!-- # clean up large objects -->
<!-- seur <- NULL -->
<!-- gc() -->
<!-- ``` -->

# CellChat: All Groups
[CellChat^1^](http://www.cellchat.org/) v2.1.1 is an R package that is able to infer cell-cell communication using single-cell RNA-seq data.  The gene expression is compared to a CellChat curated database of known ligand, receptor, and cofactor interactions.  The CellChat v2 database contains ~3,300 validated molecular interactions.  

CellChat processing starts by identifying over expressed genes and interactions and projecting that information onto a Protein-Protein Interaction network (PPI).  Biologically significant cell-cell communication interactions are assigned a probability value by modeling gene expression normalized using a method called trimean with prior knowledge of interactions using the law of mass action.  Once probability has been calculated permutation testing is performed.  Cell-cell communications were filtered out if there were fewer than 10 cells in the comparison.  The parameter `population.size = TRUE` was used for this calculation meaning that CellChat took into account that this data is a population of cells as opposed to a subset of cell-sorted cells.

CellChat analysis was performed on <font color=`r colors[params$timepoint]`> **trimester `r params$timepoint`**</font> HBV-infected subjects that experienced a flare **before** birth, **after** birth, HBV-infected subjects **without** a flare, and **uninfected** subjects.  The overall section shows a comparison of all interactions and counts in a histogram and cell-cell communication for each flare group before, after, without, and uninfected.


```{r create cellchat object}
# CellChat object created via another script "01_installCellChat_split_Bcells.R"
# Combine all 4 objects to increase power https://github.com/jinworks/CellChat/issues/62
cellchat.control <- readRDS(Sys.glob(file.path(cellchat_path, paste0("*", params$timepoint, "_control_cellchat.rds"))))
cellchat.before <- readRDS(Sys.glob(file.path(cellchat_path, paste0("*", params$timepoint, "_before_cellchat.rds"))))
cellchat.after <- readRDS(Sys.glob(file.path(cellchat_path, paste0("*", params$timepoint, "_after_cellchat.rds"))))
cellchat.without <- readRDS(Sys.glob(file.path(cellchat_path, paste0("*", params$timepoint, "_without_cellchat.rds"))))

object.list <- list(after = cellchat.after, before = cellchat.before, control = cellchat.control, without = cellchat.without)

cellchat <- readRDS(Sys.glob(file.path(cellchat_path, paste0("*", params$timepoint, "_all_cellchat.rds"))))
```

## Cell Type Count
For reference, a table that contains the number of cells in each cell type by sample is provided below. The cell types are not evenly distributed across each sample and may play a contributing factor to the differences in interaction counts and weights detected.

```{r cell count}
#data <- do.call(rbind, sapply(names(cellchat@idents), function(x) cbind(x, as.character(cellchat@idents[[x]])))) %>% as.data.frame() %>% table() %>% as.data.frame.matrix() 
data <- do.call(rbind, names(cellchat@idents) %>% setdiff(y='joint') %>% sapply( function(x) cbind(x, as.character(cellchat@idents[[x]])))) %>% as.data.frame() %>% table() %>% as.data.frame.matrix
#colnames(data) <- c("Group", "Predicted Cell Type")
htmltools::tagList(DT::datatable(data, extensions='Buttons', options=list(dom='lBfrtip', buttons = c('copy', 'csv', 'excel'), lengthMenu = list(c(10,25,50,-1), c(10,25,50,"All")))))
```

## Interaction Counts and Weights {.tabset .tabset-fade}
For the *histogram tab* the histogram on the left contains the interaction count for each group of the flare groups.  The histogram on the right contains the total interaction strength for each group.  The interaction strength is the communication probability mentioned above.

For the flare group tabs the circos plot shows the interactions between the different cell types looking only at the given flare group.  Each cell type is represented as a node with a unique color and the interaction as a line (edge) connecting cell types. At the end of each edge is an arrow indicating the direction of the interaction.  

The *Interactions tab* will display a circos plot on the left denoting interaction counts between cell types and the circos plot on the right denoting interaction strength between cell types.  

The *Weight Split on Cell Type tab* splits the interaction strength circos plot in the Interactions tab on cell type.  

### Histogram {-}
```{r Interaction Count and Weights Histogram, fig.width=12}
gg1 <- compareInteractions(cellchat, show.legend = F, group = c(1:4))
gg2 <- compareInteractions(cellchat, show.legend = F, group = c(1:4), measure = "weight")
gg1 + gg2
```

### Uninfected {.tabset .tabset-fade .tabset-pills -}
#### Interactions {-}

```{r uninfected interactions}
plotCellChat <- function(cellchat, flare){
  groupSize <- as.numeric(table(cellchat@idents))
  par(mfrow = c(1,2), xpd=TRUE)
  netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = paste0("Number of interactions: ", flare))
  netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = paste0("Interaction weights/strength: ", flare))
}

plotCellChat(cellchat.control, "uninfected")
```

#### Weight Split on Cell Type {-}
```{r uninfected Weight Split on Cell-Type}
mat <- cellchat.control@net$weight
groupSize <- as.numeric(table(cellchat.control@idents))
par(mfrow = c(3,3), xpd=TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
```

### Before {.tabset .tabset-fade .tabset-pills -}
#### Interactions {-}
```{r before interactions}
plotCellChat(cellchat.before, "before")
```

#### Weight Split on Cell-Type {-}
```{r before Weight Split on Cell-Type}
mat <- cellchat.before@net$weight
groupSize <- as.numeric(table(cellchat.before@idents))
par(mfrow = c(3,3), xpd=TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
```

### After {.tabset .tabset-fade .tabset-pills -}
#### Interactions {-}
```{r after interactions}
plotCellChat(cellchat.after, "after")
```

#### Weight Split on Cell-Type {-}
```{r after Weight Split on Cell-Type}
mat <- cellchat.after@net$weight
groupSize <- as.numeric(table(cellchat.after@idents))
par(mfrow = c(3,3), xpd=TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
```

### Without {.tabset .tabset-fade .tabset-pills -}
#### Interactions {-}
```{r without interactions}
plotCellChat(cellchat.without, "without")
```

#### Weight Split on Cell-Type {-}
```{r without Weight Split on Cell-Type}
mat <- cellchat.without@net$weight
groupSize <- as.numeric(table(cellchat.without@idents))
par(mfrow = c(3,3), xpd=TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
```

```{r Interaction caption preparation, include=FALSE}
number <- netVisual_heatmap(cellchat, comparison = comparison)
strength <- netVisual_heatmap(cellchat, measure = "weight", comparison = comparison)
strength_max <- strength@matrix == max(strength@matrix, na.rm=T)
number_max <- number@matrix == max(number@matrix, na.rm=T)

number@matrix[is.na(number@matrix)] <- 0
strength@matrix[is.na(strength@matrix)] <- 0

strength_num_diff <- strength@matrix == min(strength@matrix[number@matrix > 0 & strength@matrix < 0], na.rm=T)

label <- if (str_to_lower(group) != 'uninfected') {print(paste0('HBV-subjects that experienced a flare ', group, ' birth'))} else {print(paste0(str_to_lower(group), ' subjects'))}

```
# CellChat: Without vs `r str_to_title(params$group)`
CellChat differential analysis compared <font color=`r colors[params$timepoint]`>**trimester `r params$timepoint`**</font> HBV-infected subjects without a flare and `r label`.

## Interaction Counts and Weights {.tabset .tabset-fade}
Plots in this section will have the term differential indicating a subtraction between groups. The color <span style="color: red;">red</span> in the plots indicates <span style="color: red;">increased</span> signaling in the <span style="color: red;">without</span> group compared to the `r params$group` group and the color <span style="color: blue;">blue</span> indicates <span style="color: blue;">decreased</span> signaling in the <span style="color: blue;">without</span> group compared to the `r params$group` group.

### Relative Circos Plot Comparison {.tabset .tabset-fade .tabset-pills -}
The circos plot on the left shows the differential interaction counts between each cell type and on the right the differential interaction strength between each cell type.  The interaction lines (edges) have an arrow next to each cell type indicating the direction of the interaction.  

As an example when looking at the differential interaction strength circos plot on the right there is a red line connecting the `r names(which(rowSums(strength@matrix == max(strength@matrix), na.rm=T) > 0))` to `r names(which(colSums(strength@matrix == max(strength@matrix), na.rm=T) > 0))`.  This indicates that the interaction strength is **increased** in the **without** group compared to the `r params$group` group.  The line width indicates the strength of the difference.  

#### Without Edge Weights {-}
```{r Interaction Count and Weights Circlize, fig.width=12, fig.height=6}
par(mfrow = c(1,2), xpd=TRUE)
netVisual_diffInteraction(cellchat, weight.scale = T, comparison = comparison, title.name = )
netVisual_diffInteraction(cellchat, weight.scale = T, measure = "weight", comparison = comparison)
```

#### With Edge Weights {-}
```{r Interaction Count and Weights Circlize Edge Weights, fig.width=12, fig.height=6}
par(mfrow = c(1,2), xpd=TRUE)
netVisual_diffInteraction(cellchat, weight.scale = T, comparison = comparison, label.edge = T, title.name = )
netVisual_diffInteraction_strength(cellchat, weight.scale = T, measure = "weight", label.edge = T, comparison = comparison)
```

### Relative Interaction Counts Heatmap {-}
Interaction counts and weights seen in the circos plots are now shown as a heatmap.  The left heatmap is differential number of interactions and the right heatmap the differential interaction strength.  The cell type sending the  signal is on the y-axis and the cell type receiving the signal is on the x-axis.  

The histogram on the top of the plots shows the absolute sum of the incoming signal and the histogram on the right shows the absolute sum of the outgoing signal for each cell type.

As an example when `r names(which(rowSums(strength_num_diff, na.rm=T) > 0))` are the sender and `r names(which(colSums(strength_num_diff, na.rm=T) > 0))` are the receiver in the left heatmap the heatmap is red indicating an increase in the number of interactions in the without group.  If you then look at the right heatmap for the same comparison the heatmap is blue indicating a decrease in signaling strength in the without group.  This is an example that shows that increased signaling does not equal increased strength.

```{r Interaction Count and Weights Heatmap, warning=FALSE, fig.width=12}
gg1 <- netVisual_heatmap(cellchat, comparison = comparison)
gg2 <- netVisual_heatmap(cellchat, measure = "weight", comparison = comparison)
gg1 + gg2
```

## Sources and Targets {.tabset .tabset-fade}
```{r Sources and targets preparation, include=FALSE}
celltypes <- cellchat@meta$predicted.celltype.l1.split %>% unique %>% grep(pattern='other', invert=T, value=T) %>% sort()

sapply(celltypes, function(x) netAnalysis_signalingChanges_scatter(cellchat, idents.use = x, comparison = comparison))
temp <- sapply(celltypes, function(x) netAnalysis_signalingChanges_scatter(cellchat, idents.use = x, comparison = comparison))['data',]
for (i in names(temp)) {
  temp[[i]]$celltype <- i
}
all_changes <- do.call(rbind, temp)
specific <- all_changes$specificity %>% unique %>% grep(pattern='specific', value=T)
filtered <- all_changes %>% filter(outgoing > 0 & incoming > 0) %>% filter(specificity %in% specific) %>% filter(specificity.out.in == 'Outgoing specific')
example <- rownames(filtered)[which.max(filtered$outgoing + filtered$incoming)]

specific_color <- list(Shared = 'black', `without specific` = 'blue')
specific_color[specific %>% grep(pattern='without', invert=T, value=T)] <- 'red'

specific_shape <- list(Shared = 'circle', `Incoming specific` = 'square', `Outgoing specific` = 'triangle', `Incoming & Outgoing specific` = 'diamond')

```

The *Full Comparison tab* shows outgoing and incoming interaction strength comparison for <font color=`r colors[params$timepoint]`>**trimester `r params$timepoint`**</font> `r params$group` (left) and without (right) flare groupings for each cell type.

The cell type specific tabs show the differential signaling changes between `r params$group` and without.  The plot is shown as a coordinate plane with 4 quadrants as shown in the image below.  Points in quadrant 1 will indicate increased incoming and outgoing signaling strength in the without group. Points in quadrant 3 will indicate increased incoming and outgoing signaling strength in the `r params$group` group.  

```{r quadrants, fig.align="center"}
include_graphics("/data/NHLBI_IDSS/rawdata/NIDDK-15_temp/cellchat_BD/CellChatQuadrantBackground.png")
```

The point colors indicate if a signal was seen in both the `r params$group` and without group, the `r params$group` group only, or the without group only.  The shapes indicate whether the signal was seen in incoming and outgoing for both the `r params$group` and without group (shared), incoming specific (for `r params$group` or without, color will indicate which), outgoing specific (for `r params$group` or without, color will indicate which), and incoming & outgoing specific (e.g. incoming in the `r params$group` group only and outgoing in the without group only).

The axes are used to denote a difference in signaling strength, while the shape and the color denotes where the signal is being detected. A point / dot on the graph that is circular and black means that the signal is detected in the incoming and outgoing direction for both groups. However, this does not mean how much of a difference in these signal strengths are detected. It is possible for the signal strength to only differ in the incoming or the outgoing direction, which would result in a circle dot being on one of the axes. This is the same for points that represent signals that have an incoming specific or outgoing specific signal. If there is an incoming specific signal (square) then we would expect a non-zero difference along the incoming (y) axis, but that doesn’t mean that there is not a difference in the outgoing signal strength that would result in the point having a non-zero value along the outgoing (x) axis. The opposite is true for outgoing specific signals (triangle). Incoming and outgoing specific points have a diamond shape, and they represent signals that are only detected in one of the groups in both the incoming and outgoing direction. 

For example if you look at the `r all_changes[example, 'celltype']` plot at the top of the plot in quadrant 1 will be a `r specific_color[[as.character(all_changes[example, 'specificity'])]]` `r specific_shape[[as.character(all_changes[example, 'specificity.out.in'])]]` labeled `r all_changes[example, 'labels']`.  The location of `r all_changes[example, 'labels']` in quadrant 1 of the coordinate plane of the plot indicates that incoming and outgoing signaling strength is higher in the without group.  The `r specific_shape[[as.character(all_changes[example, 'specificity.out.in'])]]` indicates `r all_changes[example, 'specificity.out.in'] %>% str_to_lower()` and the color indicates `r all_changes[example, 'specificity'] %>% str_to_lower()`.  This means that the `r all_changes[example, 'specificity.out.in'] %>% str_to_lower() %>% gsub(pattern=' specific', replacement='')` signal was seen only in the `r all_changes[example, 'specificity'] %>% str_to_lower() %>% gsub(pattern=' specific', replacement='')` group.  Points other than black circles indicate that a signal was not present/detected in both groups.

### Full Comparison {-}
```{r Interaction Comparison Sources and Targets, fig.width=17}
num.link <- sapply(object.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
weight.MinMax <- c(min(num.link), max(num.link)) # control the dot size in the different datasets
gg <- list()
for (i in comparison) {
  x <- match(i, comparison)
  gg[[x]] <- netAnalysis_signalingRole_scatter(object.list[[i]], title = names(object.list)[i], weight.MinMax = weight.MinMax)
}

axis_min <- sapply(gg, function(x) min(layer_scales(x)$x$range$range[1], layer_scales(x)$y$range$range[1])) %>% min
axis_max <- sapply(gg, function(x) max(layer_scales(x)$x$range$range[2], layer_scales(x)$y$range$range[2])) %>% max

#> Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
#> Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
#patchwork::wrap_plots(plots = gg)
patchwork::wrap_plots(plots=lapply(gg, function(x) x + xlim(axis_min, axis_max) + ylim(axis_min, axis_max)))
```

```{r Interaction Comparison Cell Types, results='asis'}
for (celltype in celltypes) {
  cat("### ", celltype, "{-}\n\n")
  print(netAnalysis_signalingChanges_scatter(cellchat, idents.use = celltype, comparison = comparison))
  cat("\n\n")
}

```

# Session Info
```{r Session Info}
sessionInfo()
```

# References
###### 1. Suoqin Jin et al., Inference and analysis of cell-cell communication using CellChat. Nature Communications, 12:1088 (2021). https://www.nature.com/articles/s41467-021-21246-9 {-}


