---
title: "NIDDK-15 CellChat Analysis"
subtitle: "Trimester `r params$timepoint`"
author: "Vicky Chen"
date: "`r Sys.Date()`"
output:
    rmdformats::robobook:
        lightbox: true
        number_sections: true
        code-fold: true
        toc_depth: 4
        fig_width: 15
        fig_height: 8
params:
  timepoints: timepoint
  group: group
  path: path
---

```{r setup, include=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = FALSE, include=TRUE, message=FALSE, warning=FALSE, error=FALSE)
```

```{r libraries}
library(CellChat)
library(Seurat)
library(patchwork)
library(ComplexHeatmap)
library(DT)
library(knitr)
library(stringr)
options(stringsAsFactors = FALSE)
```

```{r paths}
seurat_obj <- "/data/NHLBI_IDSS/rawdata/NIDDK-15_temp/count/merge_processing/harmony/timepoint_patient/analysis1/analysis1_subset.rds"
# rename B-cells
b_cells <- read.csv("/data/NHLBI_IDSS/rawdata/NIDDK-15_temp/count/merge_processing/seurat_merge/B_cells/Bcells_1Tri_FCRL1_metadata.csv", row.names = 1)

index <- list(after=1, before=2, uninfected=3, without=4)
comparison <- c(index[[params$group]], 4) # 1 = after, 2 = before, 3 = uninfected, 4 = without 
```

```{r settings}
colors = list(`1`='#337AB7', `2`='#6AA84F', `3`='#A70505')
```

```{r redefine function}
netVisual_circle_strength <- function (net, color.use = NULL, title.name = NULL, sources.use = NULL,
    targets.use = NULL, idents.use = NULL, remove.isolate = FALSE,
    top = 1, weight.scale = FALSE, vertex.weight = 20, vertex.weight.max = NULL,
    vertex.size.max = NULL, vertex.label.cex = 1, vertex.label.color = "black",
    edge.weight.max = NULL, edge.width.max = 8, alpha.edge = 0.6,
    label.edge = FALSE, edge.label.color = "black", edge.label.cex = 0.8,
    edge.curved = 0.2, shape = "circle", layout = in_circle(),
    margin = 0.2, vertex.size = NULL, arrow.width = 1, arrow.size = 0.2,
    text.x = 0, text.y = 1.5)
{
    if (!is.null(vertex.size)) {
        warning("'vertex.size' is deprecated. Use `vertex.weight`")
    }
    if (is.null(vertex.size.max)) {
        if (length(unique(vertex.weight)) == 1) {
            vertex.size.max <- 5
        }
        else {
            vertex.size.max <- 15
        }
    }
    options(warn = -1)
    thresh <- stats::quantile(net, probs = 1 - top)
    net[net < thresh] <- 0
    if ((!is.null(sources.use)) | (!is.null(targets.use)) | (!is.null(idents.use))) {
        if (is.null(rownames(net))) {
            stop("The input weighted matrix should have rownames!")
        }
        cells.level <- rownames(net)
        df.net <- reshape2::melt(net, value.name = "value")
        colnames(df.net)[1:2] <- c("source", "target")
        if (!is.null(sources.use)) {
            if (is.numeric(sources.use)) {
                sources.use <- cells.level[sources.use]
            }
            df.net <- subset(df.net, source %in% sources.use)
        }
        if (!is.null(targets.use)) {
            if (is.numeric(targets.use)) {
                targets.use <- cells.level[targets.use]
            }
            df.net <- subset(df.net, target %in% targets.use)
        }
        if (!is.null(idents.use)) {
            if (is.numeric(idents.use)) {
                idents.use <- cells.level[idents.use]
            }
            df.net <- filter(df.net, (source %in% idents.use) |
                (target %in% idents.use))
        }
        df.net$source <- factor(df.net$source, levels = cells.level)
        df.net$target <- factor(df.net$target, levels = cells.level)
        df.net$value[is.na(df.net$value)] <- 0
        net <- tapply(df.net[["value"]], list(df.net[["source"]],
            df.net[["target"]]), sum)
    }
    net[is.na(net)] <- 0
    if (remove.isolate) {
        idx1 <- which(Matrix::rowSums(net) == 0)
        idx2 <- which(Matrix::colSums(net) == 0)
        idx <- intersect(idx1, idx2)
        net <- net[-idx, ]
        net <- net[, -idx]
    }
    g <- graph_from_adjacency_matrix(net, mode = "directed",
        weighted = T)
    edge.start <- igraph::ends(g, es = igraph::E(g), names = FALSE)
    coords <- layout_(g, layout)
    if (nrow(coords) != 1) {
        coords_scale = scale(coords)
    }
    else {
        coords_scale <- coords
    }
    if (is.null(color.use)) {
        color.use = scPalette(length(igraph::V(g)))
    }
    if (is.null(vertex.weight.max)) {
        vertex.weight.max <- max(vertex.weight)
    }
    vertex.weight <- vertex.weight/vertex.weight.max * vertex.size.max +
        5
    loop.angle <- ifelse(coords_scale[igraph::V(g), 1] > 0, -atan(coords_scale[igraph::V(g),
        2]/coords_scale[igraph::V(g), 1]), pi - atan(coords_scale[igraph::V(g),
        2]/coords_scale[igraph::V(g), 1]))
    igraph::V(g)$size <- vertex.weight
    igraph::V(g)$color <- color.use[igraph::V(g)]
    igraph::V(g)$frame.color <- color.use[igraph::V(g)]
    igraph::V(g)$label.color <- vertex.label.color
    igraph::V(g)$label.cex <- vertex.label.cex
    if (label.edge) {
        igraph::E(g)$label <- igraph::E(g)$weight
        igraph::E(g)$label <- round(igraph::E(g)$label, digits = 2)
    }
    if (is.null(edge.weight.max)) {
        edge.weight.max <- max(igraph::E(g)$weight)
    }
    if (weight.scale == TRUE) {
        igraph::E(g)$width <- 0.3 + igraph::E(g)$weight/edge.weight.max *
            edge.width.max
    }
    else {
        igraph::E(g)$width <- 0.3 + edge.width.max * igraph::E(g)$weight
    }
    igraph::E(g)$arrow.width <- arrow.width
    igraph::E(g)$arrow.size <- arrow.size
    igraph::E(g)$label.color <- edge.label.color
    igraph::E(g)$label.cex <- edge.label.cex
    igraph::E(g)$color <- grDevices::adjustcolor(igraph::V(g)$color[edge.start[,
        1]], alpha.edge)
    igraph::E(g)$loop.angle <- rep(0, length(igraph::E(g)))
    if (sum(edge.start[, 2] == edge.start[, 1]) != 0) {
        igraph::E(g)$loop.angle[which(edge.start[, 2] == edge.start[,
            1])] <- loop.angle[edge.start[which(edge.start[,
            2] == edge.start[, 1]), 1]]
    }
    radian.rescale <- function(x, start = 0, direction = 1) {
        c.rotate <- function(x) (x + start)%%(2 * pi) * direction
        c.rotate(scales::rescale(x, c(0, 2 * pi), range(x)))
    }
    label.locs <- radian.rescale(x = 1:length(igraph::V(g)),
        direction = -1, start = 0)
    label.dist <- vertex.weight/max(vertex.weight) + 2
    plot(g, edge.curved = edge.curved, vertex.shape = shape,
        layout = coords_scale, margin = margin, vertex.label.dist = label.dist,
        vertex.label.degree = label.locs, vertex.label.family = "Helvetica",
        edge.label.family = "Helvetica")
    if (!is.null(title.name)) {
        text(text.x, text.y, title.name, cex = 1.1)
    }
    gg <- recordPlot()
    return(gg)
}

netVisual_diffInteraction_strength <- function (object, comparison = c(1, 2), measure = c("count",
    "weight", "count.merged", "weight.merged"), color.use = NULL,
    color.edge = c("#b2182b", "#2166ac"), title.name = NULL,
    sources.use = NULL, targets.use = NULL, remove.isolate = FALSE,
    top = 1, weight.scale = FALSE, vertex.weight = 20, vertex.weight.max = NULL,
    vertex.size.max = 15, vertex.label.cex = 1, vertex.label.color = "black",
    edge.weight.max = NULL, edge.width.max = 8, alpha.edge = 0.6,
    label.edge = FALSE, edge.label.color = "black", edge.label.cex = 0.8,
    edge.curved = 0.2, shape = "circle", layout = in_circle(),
    margin = 0.2, arrow.width = 1, arrow.size = 0.2)
{
    options(warn = -1)
    measure <- match.arg(measure)
    obj1 <- object@net[[comparison[1]]][[measure]]
    obj2 <- object@net[[comparison[2]]][[measure]]
    net.diff <- obj2 - obj1
    if (measure %in% c("count", "count.merged")) {
        if (is.null(title.name)) {
            title.name = "Differential number of interactions"
        }
    }
    else if (measure %in% c("weight", "weight.merged")) {
        if (is.null(title.name)) {
            title.name = "Differential interaction strength"
        }
    }
    net <- net.diff
    if ((!is.null(sources.use)) | (!is.null(targets.use))) {
        df.net <- reshape2::melt(net, value.name = "value")
        colnames(df.net)[1:2] <- c("source", "target")
        if (!is.null(sources.use)) {
            if (is.numeric(sources.use)) {
                sources.use <- rownames(net.diff)[sources.use]
            }
            df.net <- subset(df.net, source %in% sources.use)
        }
        if (!is.null(targets.use)) {
            if (is.numeric(targets.use)) {
                targets.use <- rownames(net.diff)[targets.use]
            }
            df.net <- subset(df.net, target %in% targets.use)
        }
        cells.level <- rownames(net.diff)
        df.net$source <- factor(df.net$source, levels = cells.level)
        df.net$target <- factor(df.net$target, levels = cells.level)
        df.net$value[is.na(df.net$value)] <- 0
        net <- tapply(df.net[["value"]], list(df.net[["source"]],
            df.net[["target"]]), sum)
        net[is.na(net)] <- 0
    }
    if (remove.isolate) {
        idx1 <- which(Matrix::rowSums(net) == 0)
        idx2 <- which(Matrix::colSums(net) == 0)
        idx <- intersect(idx1, idx2)
        net <- net[-idx, ]
        net <- net[, -idx]
    }
    net[abs(net) < stats::quantile(abs(net), probs = 1 - top,
        na.rm = T)] <- 0
    g <- graph_from_adjacency_matrix(net, mode = "directed",
        weighted = T)
    edge.start <- igraph::ends(g, es = igraph::E(g), names = FALSE)
    coords <- layout_(g, layout)
    if (nrow(coords) != 1) {
        coords_scale = scale(coords)
    }
    else {
        coords_scale <- coords
    }
    if (is.null(color.use)) {
        color.use = scPalette(length(igraph::V(g)))
    }
    if (is.null(vertex.weight.max)) {
        vertex.weight.max <- max(vertex.weight)
    }
    vertex.weight <- vertex.weight/vertex.weight.max * vertex.size.max +
        5
    loop.angle <- ifelse(coords_scale[igraph::V(g), 1] > 0, -atan(coords_scale[igraph::V(g),
        2]/coords_scale[igraph::V(g), 1]), pi - atan(coords_scale[igraph::V(g),
        2]/coords_scale[igraph::V(g), 1]))
    igraph::V(g)$size <- vertex.weight
    igraph::V(g)$color <- color.use[igraph::V(g)]
    igraph::V(g)$frame.color <- color.use[igraph::V(g)]
    igraph::V(g)$label.color <- vertex.label.color
    igraph::V(g)$label.cex <- vertex.label.cex
    if (label.edge) {
        igraph::E(g)$label <- igraph::E(g)$weight
        igraph::E(g)$label <- round(igraph::E(g)$label, digits = 2)
    }
    igraph::E(g)$arrow.width <- arrow.width
    igraph::E(g)$arrow.size <- arrow.size
    igraph::E(g)$label.color <- edge.label.color
    igraph::E(g)$label.cex <- edge.label.cex
    igraph::E(g)$color <- ifelse(igraph::E(g)$weight > 0, color.edge[1],
        color.edge[2])
    igraph::E(g)$color <- grDevices::adjustcolor(igraph::E(g)$color,
        alpha.edge)
    igraph::E(g)$weight <- abs(igraph::E(g)$weight)
    if (is.null(edge.weight.max)) {
        edge.weight.max <- max(igraph::E(g)$weight)
    }
    if (weight.scale == TRUE) {
        igraph::E(g)$width <- 0.3 + igraph::E(g)$weight/edge.weight.max *
            edge.width.max
    }
    else {
        igraph::E(g)$width <- 0.3 + edge.width.max * igraph::E(g)$weight
    }
    igraph::E(g)$loop.angle <- 0
    if (sum(edge.start[, 2] == edge.start[, 1]) != 0) {
        igraph::E(g)$loop.angle[which(edge.start[, 2] == edge.start[,
            1])] <- loop.angle[edge.start[which(edge.start[,
            2] == edge.start[, 1]), 1]]
    }
    radian.rescale <- function(x, start = 0, direction = 1) {
        c.rotate <- function(x) (x + start)%%(2 * pi) * direction
        c.rotate(scales::rescale(x, c(0, 2 * pi), range(x)))
    }
    label.locs <- radian.rescale(x = 1:length(igraph::V(g)),
        direction = -1, start = 0)
    label.dist <- vertex.weight/max(vertex.weight) + 2
    plot(g, edge.curved = edge.curved, vertex.shape = shape,
        layout = coords_scale, margin = margin, vertex.label.dist = label.dist,
        vertex.label.degree = label.locs, vertex.label.family = "Helvetica",
        edge.label.family = "Helvetica")
    if (!is.null(title.name)) {
        text(0, 1.5, title.name, cex = 1.1)
    }
    gg <- recordPlot()
    return(gg)
}

netVisual_circle_edit <- function (net, color.use = NULL, title.name = NULL, sources.use = NULL,
          targets.use = NULL, idents.use = NULL, remove.isolate = FALSE,
          top = 1, weight.scale = FALSE, vertex.weight = 20, vertex.weight.max = NULL,
          vertex.size.max = NULL, vertex.label.cex = 1, vertex.label.color = "black",
          edge.weight.max = NULL, edge.width.max = 8, alpha.edge = 0.6,
          label.edge = FALSE, edge.label.color = "black", edge.label.cex = 0.8,
          edge.curved = 0.2, shape = "circle", layout = in_circle(),
          margin = 0.2, vertex.size = NULL, arrow.width = 1, arrow.size = 0.2,
          text.x = 0, text.y = 1.5, labels=NULL, label.cex=1)
{
  if (!is.null(vertex.size)) {
    warning("'vertex.size' is deprecated. Use `vertex.weight`")
  }
  if (is.null(vertex.size.max)) {
    if (length(unique(vertex.weight)) == 1) {
      vertex.size.max <- 5
    }
    else {
      vertex.size.max <- 15
    }
  }
  options(warn = -1)
  thresh <- stats::quantile(net, probs = 1 - top)
  net[net < thresh] <- 0
  if ((!is.null(sources.use)) | (!is.null(targets.use)) | (!is.null(idents.use))) {
    if (is.null(rownames(net))) {
      stop("The input weighted matrix should have rownames!")
    }
    cells.level <- rownames(net)
    df.net <- reshape2::melt(net, value.name = "value")
    colnames(df.net)[1:2] <- c("source", "target")
    if (!is.null(sources.use)) {
      if (is.numeric(sources.use)) {
        sources.use <- cells.level[sources.use]
      }
      df.net <- subset(df.net, source %in% sources.use)
    }
    if (!is.null(targets.use)) {
      if (is.numeric(targets.use)) {
        targets.use <- cells.level[targets.use]
      }
      df.net <- subset(df.net, target %in% targets.use)
    }
    if (!is.null(idents.use)) {
      if (is.numeric(idents.use)) {
        idents.use <- cells.level[idents.use]
      }
      df.net <- filter(df.net, (source %in% idents.use) |
                         (target %in% idents.use))
    }
    df.net$source <- factor(df.net$source, levels = cells.level)
    df.net$target <- factor(df.net$target, levels = cells.level)
    df.net$value[is.na(df.net$value)] <- 0
    net <- tapply(df.net[["value"]], list(df.net[["source"]],
                                          df.net[["target"]]), sum)
  }
  net[is.na(net)] <- 0
  if (remove.isolate) {
    idx1 <- which(Matrix::rowSums(net) == 0)
    idx2 <- which(Matrix::colSums(net) == 0)
    idx <- intersect(idx1, idx2)
    net <- net[-idx, ]
    net <- net[, -idx]
  }
  g <- graph_from_adjacency_matrix(net, mode = "directed",
                                   weighted = T)
  edge.start <- igraph::ends(g, es = igraph::E(g), names = FALSE)
  coords <- layout_(g, layout)
  if (nrow(coords) != 1) {
    coords_scale = scale(coords)
  }
  else {
    coords_scale <- coords
  }
  if (is.null(color.use)) {
    color.use = scPalette(length(igraph::V(g)))
  }
  if (is.null(vertex.weight.max)) {
    vertex.weight.max <- max(vertex.weight)
  }
  vertex.weight <- vertex.weight/vertex.weight.max * vertex.size.max +
    5
  loop.angle <- ifelse(coords_scale[igraph::V(g), 1] > 0, -atan(coords_scale[igraph::V(g),
                                                                             2]/coords_scale[igraph::V(g), 1]), pi - atan(coords_scale[igraph::V(g),
                                                                                                                                       2]/coords_scale[igraph::V(g), 1]))
  igraph::V(g)$size <- vertex.weight
  igraph::V(g)$color <- color.use[igraph::V(g)]
  igraph::V(g)$frame.color <- color.use[igraph::V(g)]
  igraph::V(g)$label.color <- vertex.label.color
  igraph::V(g)$label.cex <- vertex.label.cex
  if (label.edge) {
    igraph::E(g)$label <- igraph::E(g)$weight
    igraph::E(g)$label <- round(igraph::E(g)$label, digits = 1)
  }
  if (is.null(edge.weight.max)) {
    edge.weight.max <- max(igraph::E(g)$weight)
  }
  if (weight.scale == TRUE) {
    igraph::E(g)$width <- 0.3 + igraph::E(g)$weight/edge.weight.max *
      edge.width.max
  }
  else {
    igraph::E(g)$width <- 0.3 + edge.width.max * igraph::E(g)$weight
  }
  igraph::E(g)$arrow.width <- arrow.width
  igraph::E(g)$arrow.size <- arrow.size
  igraph::E(g)$label.color <- edge.label.color
  igraph::E(g)$label.cex <- edge.label.cex
  igraph::E(g)$color <- grDevices::adjustcolor(igraph::V(g)$color[edge.start[,
                                                                             1]], alpha.edge)
  igraph::E(g)$loop.angle <- rep(0, length(igraph::E(g)))
  if (sum(edge.start[, 2] == edge.start[, 1]) != 0) {
    igraph::E(g)$loop.angle[which(edge.start[, 2] == edge.start[,
                                                                1])] <- loop.angle[edge.start[which(edge.start[,
                                                                                                               2] == edge.start[, 1]), 1]]
  }
  radian.rescale <- function(x, start = 0, direction = 1) {
    c.rotate <- function(x) (x + start)%%(2 * pi) * direction
    c.rotate(scales::rescale(x, c(0, 2 * pi), range(x)))
  }
  label.locs <- radian.rescale(x = 1:length(igraph::V(g)),
                               direction = -1, start = 0)
  label.dist <- (vertex.weight/max(vertex.weight) + 2)*3
  plot(g, edge.curved = edge.curved, vertex.shape = shape,
       layout = coords_scale, margin = margin, vertex.label.dist = label.dist,
       vertex.label.degree = label.locs, vertex.label.family = "Helvetica",
       edge.label.family = "Helvetica", vertex.label = labels, vertex.label.cex = label.cex)
  if (!is.null(title.name)) {
    text(text.x, text.y, title.name, cex = label.cex + 0.1)
  }
  gg <- recordPlot()
  return(gg)
}

```

<!-- # Annotated UMAPs {.tabset .tabset-fade} -->
<!-- ## Cell-Type Grouped {-} -->
<!-- ```{r annotated UMAP} -->
<!-- seur <- readRDS(seurat_obj) -->

<!-- seur <- AddMetaData(seur, b_cells) -->
<!-- seur@meta.data <- seur@meta.data %>% -->
<!--   mutate(predicted.celltype.l1.split = coalesce(FCRL1, predicted.celltype.l1)) -->

<!-- # create umap of predicted.ids -->
<!-- DimPlot(seur, reduction = 'umap', group.by = "predicted.celltype.l1.split") -->
<!-- ``` -->

<!-- ## Cell-Type Split {-} -->
<!-- ```{r split annotated UMAP} -->
<!-- # create umap of split on flare -->
<!-- DimPlot(seur, reduction = 'umap', split.by = "cr.flare", group.by = "predicted.celltype.l1.split", ncol = 2) -->
<!-- ``` -->

<!-- ## Clusters {-} -->
<!-- ```{r Cluster UMAP} -->
<!-- # create umap of split on flare -->
<!-- DimPlot(seur, reduction = 'umap') -->
<!-- ``` -->

<!-- ## Cell Counts {-} -->
<!-- ```{r cell counts} -->
<!-- table(seur@meta.data$predicted.celltype.l1.split, seur@meta.data$cr.flare) -->
<!-- ``` -->

<!-- ```{r cleanup, results="hide"} -->
<!-- # clean up large objects -->
<!-- seur <- NULL -->
<!-- gc() -->
<!-- ``` -->

# CellChat: All Groups
[CellChat^1^](http://www.cellchat.org/) v2.1.1 is an R package that is able to infer cell-cell communication using single-cell RNA-seq data.  The gene expression is compared to a CellChat curated database of known ligand, receptor, and cofactor interactions.  The CellChat v2 database contains ~3,300 validated molecular interactions.  

CellChat processing starts by identifying over expressed genes and interactions and projecting that information onto a Protein-Protein Interaction network (PPI).  Biologically significant cell-cell communication interactions are assigned a probability value by modeling gene expression normalized using a method called trimean with prior knowledge of interactions using the law of mass action.  Once probability has been calculated permutation testing is performed.  Cell-cell communications were filtered out if there were fewer than 10 cells in the comparison.  The parameter `population.size = TRUE` was used for this calculation meaning that CellChat took into account that this data is a population of cells as opposed to a subset of cell-sorted cells.

```{r, results='asis'}
cat("CellChat analysis was performed on HBV-infected subjects in ")
cat(paste(sapply(params$timepoints, function(i) paste0("<font color=", colors[i], ">**trimester ", i, "**</font>")), collapse=', '))
cat(" that experienced a flare **before** birth, **after** birth, HBV-infected subjects **without** a flare, and **uninfected** subjects.  The overall section shows a comparison of all interactions and counts in a histogram and cell-cell communication for each flare group before, after, without, and uninfected.\n\n")
```

```{r define plotCellChat}
plotCellChat <- function(cellchat, flare){
  
  test <- levels(cellchat@idents)
  test <- test %>% gsub(pattern='^B cell', replacement = 'Naïve B cell')
  test <- test %>% gsub(pattern='cell cluster', replacement = parse(text="'cell\nCluster'"))
  test[test %>% grep(pattern = 'FCRL')] <- expression(atop("FCRL1"^"high", "\nNaïve B cell"))
  #test <- test %>% gsub(pattern=' ', replacement='')
  
  count <- cellchat@net$count
  weight <- cellchat@net$weight
  
  rownames(count) <- test
  colnames(count) <- test
  
  rownames(weight) <- test
  colnames(weight) <- test
  
  colors <- c("#E41A1C", "#F781BF", "#377EB8", "#4DAF4A", "#984EA3", "#F29403", "#BC9DCC", "#A65628", "#54B0E4", "#222F75", "#1B9E77")
  names(colors) <- c(grep(pattern='FCRL', x=test, value=T), grep(pattern='FCRL', x=test, value=T, invert=T))
  
  
  groupSize <- as.numeric(table(cellchat@idents))
  par(mfrow = c(1,2), xpd=TRUE)
  netVisual_circle_edit(count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = paste0("Number of interactions: ", flare), color.use=colors[colnames(weight)], labels=test, margin = 0.5, text.y=2, label.cex=1.5)
  netVisual_circle_edit(weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = paste0("Interaction weights/strength: ", flare), color.use=colors[colnames(weight)], labels=test, margin = 0.5, text.y=2, label.cex=1.5)
}
```

## Interaction Counts and Weights 
For the flare group tabs the circos plot shows the interactions between the different cell types looking only at the given flare group.  Each cell type is represented as a node with a unique color and the interaction as a line (edge) connecting cell types. At the end of each edge is an arrow indicating the direction of the interaction.  

The *Interactions tab* will display a circos plot on the left denoting interaction counts between cell types and the circos plot on the right denoting interaction strength between cell types.  

```{r create cellchat object, results='asis'}
# CellChat object created via another script "01_installCellChat_split_Bcells.R"
# Combine all 4 objects to increase power https://github.com/jinworks/CellChat/issues/62

for (timepoint in params$timepoints) {
  cellchat_path <- file.path(params$path, paste0("trimester", timepoint))
  
  cellchat.control <- readRDS(Sys.glob(file.path(cellchat_path, paste0("*", params$timepoint, "_control_cellchat.rds"))))
  cellchat.before <- readRDS(Sys.glob(file.path(cellchat_path, paste0("*", params$timepoint, "_before_cellchat.rds"))))
  cellchat.after <- readRDS(Sys.glob(file.path(cellchat_path, paste0("*", params$timepoint, "_after_cellchat.rds"))))
  cellchat.without <- readRDS(Sys.glob(file.path(cellchat_path, paste0("*", params$timepoint, "_without_cellchat.rds"))))
  
  cat(paste0("### <font color=", colors[timepoint], ">Trimester"), timepoint, "</font>{.tabset .tabset-fade .tabset-pills}\n\n")
  
  cat("#### Uninfected {.tabset .tabset-fade .tabset-pills -}\n\n")
  plotCellChat(cellchat.control, "uninfected")
  cat("\n\n")
  
  cat("#### After {.tabset .tabset-fade .tabset-pills -}\n\n")
  plotCellChat(cellchat.after, "after")
  cat("\n\n")
  
  cat("#### Before {.tabset .tabset-fade .tabset-pills -}\n\n")
  plotCellChat(cellchat.before, "before")
  cat("\n\n")
  
  cat("#### Without {.tabset .tabset-fade .tabset-pills -}\n\n")
  plotCellChat(cellchat.without, "without")
  cat("\n\n")
}
```




```{r uninfected interactions, fig.width=15}



```

<!-- #### Weight Split on Cell Type {-} -->
<!-- ```{r uninfected Weight Split on Cell-Type} -->
<!-- mat <- cellchat.control@net$weight -->
<!-- groupSize <- as.numeric(table(cellchat.control@idents)) -->
<!-- par(mfrow = c(3,3), xpd=TRUE) -->
<!-- for (i in 1:nrow(mat)) { -->
<!--   mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat)) -->
<!--   mat2[i, ] <- mat[i, ] -->
<!--   netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i]) -->
<!-- } -->
<!-- ``` -->

# Session Info
```{r Session Info}
sessionInfo()
```

# References
###### 1. Suoqin Jin et al., Inference and analysis of cell-cell communication using CellChat. Nature Communications, 12:1088 (2021). https://www.nature.com/articles/s41467-021-21246-9 {-}


